### PR TITLE
Add mBaaS configuration

### PR DESCRIPTION
**Pull request by TsvetanMilanov**
We need to add mBaaS configuration to use Kinvey. The mBaaS service will
use TFIS for authentication. That's why we need to reuse our current
server configuration and change it to work with the Kinvey service.

We need to create public nsCloudKinveyService. It must have the following
methods:
- `getApps(): Promise<IKinveyApplication[]>;` -> This method returns all
  applications of the current user.
- `createApp(input: ICreateKinveyAppInput): Promise<IKinveyApplication>;`
  -> This method creates new app for the user.
- `createAuthService(input: ICreateKinveyAuthService):
  Promise<IKinveyAuthService>;` -> this method creates identity store
  and authentication service in it. Then sets the default auth service of
  the provided environment to the newly created auth service.
- `updateAuthService(input: IUpdateKinveyAuthService):
  Promise<IKinveyAuthService>;` -> This method will update the provided
  auth service. It expects all properties of the auth service to be
  provided.
- `getAuthServices(input: IGetKinveyAuthServices):
  Promise<IDictionary<IKinveyAuthService[]>>;` -> This method returns all
  authentication services in all identity stores for the provided
  environment.
- `getDefaultAuthService(input: IGetDefaultKinveyAuthService):
  Promise<IKinveyAuthService>;` -> This method returns the default auth
  service for the provided environment.
- `changeDefaultAuthService(input: IChangeDefaultKinveyAuthService):
  Promise<IKinveyAuthService>;` -> This method changes the default auth
  service of the provided environment.

### CODE REVIEW

File: lib/services/kinvey-service.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+	public createApp(input: ICreateKinveyAppInput): Promise<IKinveyApplication> {
+		return this.$nsCloudKinveyRequestService.createApp(input);
+	}
+
+	public async createAuthService(input: ICreateKinveyAuthService): Promise<IKinveyAuthService> {
+		input.identityStoreOptions = input.identityStoreOptions || Object.create(null);
+
+		const app = await this.$nsCloudKinveyRequestService.getApp(input.appId);
+		const createIsInput: ICreateKinveyIdentityStoreInput = {
```
Comment (rosen-vladimirov) [inline lib/services/kinvey-service.ts]: `createIsInput` - I'm not sure I understand this name

File: lib/services/kinvey-service.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	public updateAuthService(input: IUpdateKinveyAuthService): Promise<IKinveyAuthService> {
+		// The Kinvey api does not accept {} for customized login.
+		if (_.isEmpty(input.authService.customizedLogin)) {
+			delete input.authService.customizedLogin;
+		}
+
+		// The Kinvey api does not accept null for provider.authentication.
+		// The api docs say it should accept null but...
```
Comment (rosen-vladimirov) [inline lib/services/kinvey-service.ts]: :smiley_cat:

File: lib/services/kinvey-service.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+			type: provider.type,
+			uniqueUserContextField: provider.uniqueUserContextField,
+			options: this.getAuthProviderOptions(provider.type, provider.options)
+		};
+
+		return result;
+	}
+
+	private getAuthProviderOptions(providerType: string, options: IKinveyAuthProviderOptions): IKinveyAuthProviderOptions {
+		if (providerType === "OAuth2") {
```
Comment (rosen-vladimirov) [inline lib/services/kinvey-service.ts]: OAuth2 and OIDC must be constants

File: Gruntfile.js [hunk 4]
```diff
@@ -138,7 +138,7 @@ module.exports = function (grunt) {
 
 			_.each(dependenciesToInstall, name => {
 				try {
-					childProcess.execSync(`npm i --ignore-scripts --production ${name}`, { cwd: pathToModule, stdio: "ignore" });
+					childProcess.execSync(`npm i --ignore-scripts ${name}`, { cwd: pathToModule, stdio: "ignore" });
```
[the remark this answers is not in the document]
Comment (TsvetanMilanov) [reply Gruntfile.js]: We get `Error: Cannot find module 'mobile-cli-lib/yok'` in our CI with `--production`.

File: xunit.xml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<testsuite name="Mocha Tests" tests="38" failures="0" errors="0" skipped="0" timestamp="Thu, 01 Mar 2018 13:50:35 GMT" time="0.19">
```
Comment (rosen-vladimirov) [inline xunit.xml]: this file should not be commited

**Pull request merged**